### PR TITLE
feat(ddm): Add support for environment filtering in metrics api

### LINKED ISSUE #59605: Add support for environment in the new metrics API

### PR DESCRIPTION
**Pull request by iambriccardo**
This PR adds the support for the environment filtering in the new metrics API.

Closes https://github.com/getsentry/sentry/issues/59605

### CODE REVIEW

File: src/sentry/sentry_metrics/querying/api.py [hunk 1]
```diff
@@ -144,9 +145,36 @@ def meta(self) -> Sequence[Mapping[str, str]]:
...
+            self._timeseries = self._timeseries.set_filters(
+                existing_filters + [Condition(Column("environment"), Op.IN, environment_names)]
+            )
+
+        return self
+
+    def alias_operators(self) -> "MutableTimeseries":
+        # In case we have a percentile in the form `px` where `x` is in the range [0-100], we want to convert it to
+        # the quantiles operation which generalizes any percentile.
+        if (match := self.PERCENTILE_REGEX.match(self._timeseries.aggregate)) is not None:
```
Comment (iambriccardo) [inline src/sentry/sentry_metrics/querying/api.py:167]: In future PRs, a more generic aliasing system will be implemented.

**Review by evanh — APPROVED** [review]

File: src/sentry/sentry_metrics/querying/api.py [hunk 2]
```diff
@@ -144,9 +145,36 @@ def meta(self) -> Sequence[Mapping[str, str]]:
...
+            existing_filters = self._timeseries.filters[:] if self._timeseries.filters else []
+            self._timeseries = self._timeseries.set_filters(
+                existing_filters + [Condition(Column("environment"), Op.IN, environment_names)]
+            )
+
+        return self
+
+    def alias_operators(self) -> "MutableTimeseries":
+        # In case we have a percentile in the form `px` where `x` is in the range [0-100], we want to convert it to
+        # the quantiles operation which generalizes any percentile.
```
Comment (jan-auer) [inline src/sentry/sentry_metrics/querying/api.py:166]: That is incorrect. We support a only pre-defined set of percentiles, hence also the hard-coded functions in the public API. There's no generic `pXX` function conversion. If we want generic percentiles, then let's expose a `percentile(X, <metric>)` function, instead. Likewise, there is no `p100` function. There is just `max`.

You could consider adding a hard-coded mapping of the four percentile functions, which also gets rid of this regex.

File: src/sentry/sentry_metrics/querying/api.py [hunk 3]
```diff
@@ -479,7 +504,7 @@ def run_metrics_query(
 
     # Parsing the input and iterating over each timeseries.
     parser = QueryParser(fields=fields, query=query, group_bys=group_bys)
-    for timeseries in parser.parse_timeserieses():
+    for timeseries in parser.generate_queries(environments=environments):
```
Comment (jan-auer) [inline src/sentry/sentry_metrics/querying/api.py:507]: This is a bit confusing now as it returns instances of `Timeseries`, but it mentions queries where an actual `Query` type exists. Would the former naming of `parse_timeseries` (see spelling) or `generate_timeseries` be better?

**Review by jan-auer — COMMENTED** [review]
Thank you, this makes sense. See two comments below that could be addressed in a follow-up.

File: src/sentry/sentry_metrics/querying/api.py (hunk 2 above)
Comment (iambriccardo) [reply src/sentry/sentry_metrics/querying/api.py:166]: We wanted to have a discussion about which percentiles we want to expose and how but the idea would be that the layer itself performs these coercions.

File: src/sentry/sentry_metrics/querying/api.py (hunk 3 above)
Comment (iambriccardo) [reply src/sentry/sentry_metrics/querying/api.py:507]: Yeah, good point

**Pull request merged**